File: leet/partition_labels_763/PartitionLabels.java
/*
LeetCode #763 Parition Labels: O(n) time O(n) space

A string S of lowercase English letters is given. We want to partition this
string into as many parts as possible so that each letter appears in at most one
part, and return a list of integers representing the size of these parts.
*/

import java.util.Map;
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public List<Integer> partitionLabels(String S) {
        Map<Character, Integer> lastPosition = new HashMap<>();
        List<Integer> output = new ArrayList<>();

        for (int i = 0; i < S.length(); i++)
            lastPosition.put(S.charAt(i), i);
        int start = 0, end = -1;
        for (int i = 0; i < S.length(); i++) {
            end = Math.max(end, lastPosition.get(S.charAt(i)));
            if (end == i) {
                output.add(end + 1 - start);
                start = end + 1;
            }
        }
        return output;
    }
}

class PartitionLabels {
    public static void main(String[] args)
    {
        String S = "ababcbacadefegdehijhklij";
        List<Integer> output = new Solution().partitionLabels(S);
        System.out.print("{");
        for (int n : output)
            System.out.print(n + ", ");
        System.out.println("}");
    }
}

File: leet/partition_labels_763/partition_labels.py
"""
LeetCode #763 Parition Labels: O(n) time O(n) space

A string S of lowercase English letters is given. We want to partition this
string into as many parts as possible so that each letter appears in at most one
part, and return a list of integers representing the size of these parts.
"""

from typing import List

class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        partitions = []
        last_position = {}
        for i, n in enumerate(S):
            last_position[n] = i
        start = 0
        end = -1
        for i, n in enumerate(S):
            end = max(end, last_position[n])
            if i == end:
                partitions.append(end + 1 - start)
                start = end + 1
        return partitions

if __name__ == "__main__":
    S = "ababcbacadefegdehijhklij"
    result = Solution().partitionLabels(S)
    print("Solution to", S, result)
    assert result == [9,7,8]
